post/2017-10-30-how-test-serverless-applications.md

```diff
@@ -0,0 +1,95 @@
+
+**Set up event mocks**<br>
+One other step that you’ll need to take before starting the tests is to have some event mocks prepared for all your handlers depending on one type of event the handler is expecting. For example, if your serverless function is subscribing to an S3 event source, make sure you have a JSON file that mocks that S3 event that AWS sends out. You can get that by trying it out yourself only once on AWS, and store it somewhere for future reference.
+
+> **Note:** It’s a known issue that each event source sends out a different event structure, and there’s no central place to see what events look like without actually trying them out yourself. We’re working behind the scenes on solving that problem. Stay tuned!
+
+## Local Lambda invocation
+
+Let's test those thin handler layers and how they fit in within your codebase. You can do so by invoking your function locally using the `serverless invoke local` command. Just provide it with the function you’d like to invoke, and an accurate event mock. (...Which you’ve totally set up already, right?!)
+
+For example, let's test a function called `createThumbnail` that is subscribing to an S3 event source. We'll do this by putting the S3 mocked event in a `createThumbnail.json` file, and then we'll run:
+
+`serverless invoke local -f createThumbnail -p createThumbnail.json`
+
+While `invoke local` doesn’t emulate Lambda 100%, you’ll still be able to find issues in your codebase quickly without having to wait for a deployment.
+
+## Using the Event Gateway
+
+We’ve recently announced a new project called the [Event Gateway](https://www.serverless.com/event-gateway) that helps manage all events happening in your serverless application. You can use the Event Gateway to locally and rapidly test your functions. 
+
+We'll need two terminal sessions for this. First, spin up the Event Gateway in your current terminal session with `serverless run`. Then, open another session and emit events with the `serverless emit` command. This will invoke any function that is subscribing to that event, and you’ll be able to see the result of all function invocations in the `serverless run` session.
+
+> **Note:** this only works for functions that are subscribed to one or more events. For other functions, you'll need to invoke the function directly, as shown above in the Local Lambda invocation section.
+
+## Remote Lambda invocation
+
+After testing and debugging your serverless application locally, you probably feel confident enough to deploy your application—at least to the dev stage. 
+
+Keep in mind that the local environment is a bit different than the actual deployment environment: e.g., AWS Lambda limits don’t apply locally, so you'll need to make sure you’re not hitting any of those limits on deployment. This is where stages come in; they can act like exact replicas of the production environment, since they run on the same infrastructure.
```

Comment:
I like the idea of this paragraph, but I feel like the great point in there is buried a bit. What about this as a restructure:

_Using multiple stages is a great way to have more confidence about your tests. You can have a QA environment that is an exact replica of your production environment, since they're running on the same infrastructure. This can ferret out hidden bugs you might miss when developing locally, such as issues with your function's IAM permissions or limitations around Lambda._

Reply:
Beautifully written! ❤️